Improved handling of Text Area focus state

diarmidmackenzie opened the pull request:
## Context

After making a pragmatic fix under PR#2263, I came across #2189, and as per comments there, I realized that there were various further issues with how focus on the text box is managed.

As described [there](https://github.com/RooVetGit/Roo-Code/issues/2189#issuecomment-2776688437), when the Roo Chat tab becomes visible again (either by switching between Roo tabs) or by switching between VSCode extensions, there is no way to determine from existing state whether or not focus should be re-instated on the text panel.

This PR explicitly tracks whether there is focus on this panel, and then re-instates it when the panel comes back into view.

## Implementation

Basic approach is:
- track focus and blur events to keep track of whether the text area has focus.
- re-instate this when the chat view reappears.
- simplify ChatView by pulling this code into a custom hook.
- revert previous fix for #2263, which is no longer required.

A bit of complexity arose because of the various ways that the text area can "blur" that we don't want to count as blurring, in particular
- switching to a new Roo tab (Prompts, MCP Servers etc.)
- switching to a different VS Code extension.

To accommodate the last of these, I had to extend the ExtensionMessage API to add a new "didBecomeInvisible" message that fires when the VSCode extension is hidden.

I added a decent set of Unit Tests to cover these various flows as well.

## Screenshots

N/A

## How to Test

Key manual tests are:
- toggle focus on the text area by clicking on / off
- focus state should be preserved when switching to a different Roo tab (e.g. Prompts, MCP Servers) and back
- focus state should be preserved when switching to a different VSCode extension so that Roo is hidden (or hiding Roo) and then switching back to Roo.

## Get in Touch

diarmid/diarmidm on Discord
<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Improves text area focus handling in `ChatView` by tracking focus state and reinstating it when the view becomes visible, using a new `useFocusPreservation` hook.
> 
>   - **Behavior**:
>     - Tracks focus and blur events in `ChatView.tsx` to manage text area focus state.
>     - Reinstates focus when the chat view becomes visible again.
>     - Introduces `useFocusPreservation` hook in `useFocusPreservation.tsx` to handle focus state management.
>     - Reverts previous fix for #2263 as it's no longer needed.
>   - **ExtensionMessage**:
>     - Adds `didBecomeInvisible` action to `ExtensionMessage` in `ExtensionMessage.ts`.
>   - **Tests**:
>     - Adds unit tests in `ChatView.test.tsx` to verify focus behavior when view visibility changes.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=RooVetGit%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for 86d9722b4ba9517a2ebfc759b969685a2455f820. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

diarmidmackenzie commented on webview-ui/src/components/chat/hooks/useFocusPreservation.tsx:
```diff
@@ -0,0 +1,70 @@
+			// Blur event should only be interpreted as an intentional defocus
+			// if they have not occured as a result of the element being hidden.
+			const BLUR_DELAY = 500
+			setTimeout(() => {
```
Agree.  Better to handle window condition with clearTimeout rather than storing / comparing timestamps.  Fixed.

libertyteeth commented:
Added the below to a closed PR ( https://github.com/RooVetGit/Roo-Code/pull/2263#issuecomment-2781345752 ); I see this one's open.  Haven't reviewed differences between the techniques; mine's a state machine approach.  Hope this helps!

---

I mentioned this in https://github.com/RooVetGit/Roo-Code/issues/2189#issuecomment-2781325938 , just wanted to note it here: I've worked on a fairly comprehensive state machine for maintaining focus.

It may be lacking a few edge cases?  I put it here a few days ago, then got busy with life :)

https://github.com/libertyteeth/roo-fixes/pull/2

diarmidmackenzie commented:
> Added the below to a closed PR ( [#2263 (comment)](https://github.com/RooVetGit/Roo-Code/pull/2263#issuecomment-2781345752) ); I see this one's open. Haven't reviewed differences between the techniques; mine's a state machine approach. Hope this helps!
> 
> I mentioned this in [#2189 (comment)](https://github.com/RooVetGit/Roo-Code/issues/2189#issuecomment-2781325938) , just wanted to note it here: I've worked on a fairly comprehensive state machine for maintaining focus.
> 
> It may be lacking a few edge cases? I put it here a few days ago, then got busy with life :)
> 
> [libertyteeth/roo-fixes#2](https://github.com/libertyteeth/roo-fixes/pull/2)

The two fixes attempt to implement different behaviours, so it depends a bit what the desired behaviour is (which I don't think is clear, and is not specified anywhere that I know, so we're both working from our intuition about what we would expect, where we might have different expectations/intuition).

The scenarios I focused on in my fix were:
- Switch to a different tab and back to chat view (isHidden true/false)
- Extension going invisible / visible.

The function I implemented was to always restore the focus on TextArea to what it was before - i.e. if it was focused before ChatView/extension was hidden, make it focused when ChatView becomes visible/not hidden again; and if it was not focused before, then don't focus it again.

This can only be achieved by actively monitoring for focus/blur events on TextArea (which I added; the existing code and your new code don't do this).

Some scenarios that I didn't consider/cover, that look to be handled in your new code:
1 - Restore of focus when Alt-Tabbing to/from the app.  On my windows system, I could not reproduce this issue
2 - Autofocus on TextArea when switching to a different task through history.  It's not obvious to me that this is desired behaviour (if it is, my code could be accommodated to cover this).
3 - Autofocus on TextArea when user clicks anywhere in the ChatWindow.

Other points to highlight from my PR, that I think are definitely beneficial.
- Pulling the focus-related logic into a separate hook, to reduce code complexity in ChatView
- Unit Tests, so we will be alerted to future regressions in this area.

If we can identify additional behaviours that are desirable, but not yet implemented in my PR (e.g. items 1-3 above, I'd be happy to extend my PR and the UTs to cover them).

hannesrudolph commented:
@diarmidmackenzie 
Based on recent updates to the focus handling (specifically, allowing messages to be entered while Roo is processing), it seems this PR is no longer applicable. I've reached out via Discord to confirm with you, but I'll close this for now to keep things tidy.

If this assessment isn't correct or there's more you'd like to discuss, please let us know and we can revisit this. For future contributions, please follow our [issue-first approach](https://github.com/RooCodeInc/Roo-Code/blob/main/CONTRIBUTING.md#issue-first-approach).

Thanks again for your effort!
